### PR TITLE
Major refactor of `FieldBoundaryConditions`

### PR DESCRIPTION
**Pull request by glwagner**
This PR introduces a major refactor of `FieldBoundaryConditions` and changes the user API.

The `grid` and field location are no longer necessary to build `FieldBoundaryConditions`. This means that the constructors `TracerBoundaryConditions`, `UVelocityBoundaryConditions`, etc, are gone. Instead we have only `FieldBoundaryConditions`, which is used like

```julia
u_bcs = FieldBoundaryConditions(top=FluxBoundaryCondition(-1e-4), bottom=FluxBoundaryCondition((x, y, t, u) = -1e-3 * u, field_dependencies=:u))

T_bcs = FieldBoundaryConditions(top=FluxBoundaryCondition(1e-8))
```

There's also a new constructor `AuxiliaryFieldBoundaryConditions` which takes the place of `ComputedFieldBoundaryConditions` and is now the default for `Field` and `ComputedField`. The difference is that `AuxiliaryFieldBoundaryConditions` have `nothing` instead of an `ImpenetrableBoundaryCondition` in `Bounded` directions when they're located on faces.

Under the hood the `FieldBoundaryConditions` struct is "flattened" so there's no more `CoordinateBoundaryConditions`. In addition, a new field `immersed` is added to `FieldBoundaryConditions` in addition to `east`, `west`, etc. This will be used to support boundary conditions on immersed boundaries in the future.

There's still quite a few validation experiments to clean up and probably some lingering issues with tests, but I was able to get the examples to run so I thought it was the right time to open the PR.

On the API changes: it seemed like a positive change that we don't need `grid` in the boundary conditions constructor, and that we only have one name `FieldBoundaryConditions` rather than 4. However, if others feel they liked the old API, we can add convenience functions `TracerBoundaryConditions`, etc back to the source.

### CODE REVIEW

**Comment by navidcy** [comment]
Ok, this definitely needs bumping up a minor release

**Comment by navidcy** [comment]
Great job. I’ll try to help out cleaning up the rest.

File: docs/src/model_setup/boundary_conditions.md [hunk 1]
```diff
@@ -238,29 +253,27 @@ When running on the GPU, `Q` must be converted to a `CuArray`.
...
+├── bottom: BoundaryCondition{Oceananigans.BoundaryConditions.Gradient, Float64}
+├── top: BoundaryCondition{Oceananigans.BoundaryConditions.Value, Int64}
+└── immersed: BoundaryCondition{Oceananigans.BoundaryConditions.Flux, Nothing}
 ```
 
-`T_bcs` is a [`FieldBoundaryConditions`](@ref) object for temperature `T` appropriate
-for horizontally periodic grid topologies.
-The default `Periodic` boundary conditions in ``x`` and ``y`` are inferred from the `topology` of `grid`.
+If the grid is horizontally-periodic, then each horizontal `DefaultPrognosticFieldBoundaryCondition`
+is converted to `PeriodicBoundaryCondition` inside the model constructor, before assigning the
```
Comment (navidcy) [inline docs/src/model_setup/boundary_conditions.md:273]:
```suggestion
If the grid is, e.g., horizontally-periodic, then each horizontal `DefaultPrognosticFieldBoundaryCondition`
is converted to `PeriodicBoundaryCondition` inside the model's constructor, before assigning the
```

**Comment by navidcy** [comment]
@glwagner, regarding the
```Julia
ERROR: LoadError: UndefVarError: bc not defined
```
in the Docs, is it related to
https://github.com/JuliaDocs/Documenter.jl/issues/228
perhaps?

**Comment by navidcy** [comment]
I believe it should be OK now. Commit e39c509 was the crucial fix here.

File: src/BoundaryConditions/field_boundary_conditions.jl [hunk 2]
```diff
@@ -73,7 +73,7 @@ Returns boundary conditions for auxiliary fields (fields
 whose values are derived from a model's prognostic fields) on `grid`
 and at `location`.
 
-Keyword arguments specify boundary conditions on the 7 possible boundaries:
+Keyword arguments specify boundary conditions on the 6 possible boundaries:
```
Comment (navidcy) [inline src/BoundaryConditions/field_boundary_conditions.jl:76]: So is this 6 or 7? I know it's the immersed also, but the list here lists only 6...

**Review by navidcy — APPROVED** [review]
I approve.

I applaud the tests/jldoctests.

I'm a bit skeptical whether I have the expertise to be as critical.

**Comment by navidcy** [comment]
https://clima.github.io/OceananigansDocumentation/previews/PR1843/

**Comment by glwagner** [comment]
> @glwagner, regarding the
> 
> ```julia
> ERROR: LoadError: UndefVarError: bc not defined
> ```
> 
> in the Docs, is it related to
> [JuliaDocs/Documenter.jl#228](https://github.com/JuliaDocs/Documenter.jl/issues/228)
> perhaps?

Nice find!

**Comment by glwagner** [comment]
I'm going to clean up the validation experiments for the new syntax and then this is ready.

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 2 above)
Comment (glwagner) [reply src/BoundaryConditions/field_boundary_conditions.jl:76]: I guess there is a field `immersed`, but it doesn't do anything yet. I think it's fine to leave this at `6`.

**Comment by glwagner** [comment]
> I approve.
> 
> I applaud the tests/jldoctests.
> 
> I'm a bit skeptical whether I have the expertise to be as critical.

If not you, then who?

**Comment by navidcy** [comment]
We should wait for this to be merged before we move on with #1847, right?

**Comment by glwagner** [comment]
When we merge this we will have to update the example in #1847 but otherwise this is just an aesthetics change and has no impact on output, so we don't need to wait for it.

**Pull request merged**